pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -158,11 +189,46 @@ func (w *HashAggFinalWorker) loadFinalResult(sctx sessionctx.Context) {
 			}
 		}
 	}
+}
 
-	w.outputCh <- &AfFinalResult{chk: result, giveBackCh: w.finalResultHolderCh}
-	if w.stats != nil {
-		w.stats.ExecTime += int64(time.Since(execStart))
+func (w *HashAggFinalWorker) loadFinalResult(sctx sessionctx.Context) {
+	waitStart := time.Now()
+	result, finished := w.receiveFinalResultHolder()
+	w.increaseWaitTime(waitStart)
+	if finished {
+		return
 	}
+
+	failpoint.Inject("ConsumeRandomPanic", nil)
+
+	execStart := time.Now()
+	defer w.increaseExecTime(execStart)
+	if w.isSpilledTriggered {
+		var ok bool
+		var err error
+		for {
+			// As we restore data by partition, only one partition uses memory at the same time.
+			// So we need to release the memory usage of last partition.
+			w.memTracker.Consume(-w.restoredAggResultMapperMem)
+			w.partialResultMap, ok, w.restoredAggResultMapperMem, err = w.getInputFromDisk(sctx)
+			if err != nil {
+				w.outputCh <- &AfFinalResult{err: err}
+				return
+			}
+
+			if !ok {
+				w.partialResultMap = nil
+				w.memTracker.Consume(-w.restoredAggResultMapperMem)
```

Comment:
Will the `w.restoredAggResultMapperMem` always be 0 when ok is not true?

Reply:
> Will the `w.restoredAggResultMapperMem` always be 0 when ok is not true?

No, in order to restore a partition, we need to read many spill files and maybe first 5 files are ok and fail to read the 6th file.